### ISSUE TITLE
Climb Score "peak detection"

### ISSUE THREAD

**Issue by Salzi** [Runalyze/Runalyze#2197]
How does the climb score evaluate if its a peak or not? Here are two examples:

![lagginhorn](https://user-images.githubusercontent.com/4997543/28559619-895899c2-7117-11e7-8128-97bb7eedbb6f.png)
![granparadiso](https://user-images.githubusercontent.com/4997543/28559621-8c8e7044-7117-11e7-92f6-2e1b9559307f.png)

Why is there a Cat. 1 peak in the upper graph (at 2.43 km) but no peak in the lower graph at ~3.2 km? I saw the same thing in different other activities too.

**Comment by mipapo** [comment]
We are using an exact epsilon (https://jsfiddle.net/75bmj0co/) (I don't know by heart which one). So there will always be some Peaks which will not count

**Comment by Salzi** [comment]
Would it make sense to make the epsilon configurable (don't know if it make sense to configure it global or by activity)?

**Comment by laufhannes** [comment]
Climb detection is done by simplifying the elevation profile by Douglas-Peucker algorithm (using shortest distance, not perpendicular distance) with an epsilon of `1.1`. I can't say why the climbs are selected as they are in your example.

Generally spoken, if you draw a line from start point to highest point, the (1) category in the first example has the largest distance to the line (which is greater than epsilon), but in the second example no point in between exceeds epsilon.

You need to fork the fiddle and add both elevation profiles to adjust the epsilon dynamically or modify the respective js file [climb-score.js#L83](https://github.com/Runalyze/Runalyze/blob/master/app/Resources/public/js/climb-score.js#L83).

Or you try do "draw" some profiles at https://karthaus.nl/rdp/ (but the epsilon is not the same as in Runalyze, as the plot uses other scales).

Making epsilon configurable requires that users can trigger a complete recalculation of all climb scores and it will result in inconsistent values among all athletes - not my favorite outcome.

**Comment by Salzi** [comment]
Thanks @laufhannes for your explanation. Is there a reason why you use the  Douglas-Peucker algorithm? As far as I understand this is just to simplify the elevation profile. Why not using some "peak finding algorithms". This are matlab scripts but maybe this could be a inspiration: https://terpconnect.umd.edu/~toh/spectrum/PeakFindingandMeasurement.htm

**Comment by laufhannes** [comment]
Generally, I'm open for a discussion about another/a better algorithm for climb detection. Douglas-Peucker was already 'there' as it's one option for elevation calculation (where it's used as line simplification as well) and in contrast to such a peak finding algorithm it results directly in start and end points for climbs, while the peak finding algorithm does return the peak only (i.e. the end point of the climb), right?

Another problem of a peak finding algorithm (don't know if that's configurable) is that they only look at peaks as 'turning points' (derivative crosses zero). I guess they have a problem with a climb to a plateau. Imagine a climb of 3k at 10%, followed by 3k at 0.5%?

**Comment by Salzi** [comment]
It really depends on what you expect and also on the sport itself. I'm more into mountaineering than running and there are definitions what a peak is (in the alps the [topographic prominence](https://en.wikipedia.org/wiki/Topographic_prominence) must be at least 30m). In your climb score you use some classification from cycling and in the end we use it for running ...
Therefore there is no solution for everything and we need a clear definitions what you want to show with the score (and yes it should be comparable to other users).

**Comment by laufhannes** [comment]
Yeah, I agree that it'd be good to have a clear definition in our glossary. From my point of view, a short draft:
> _A climb is a part of a route with positive elevation difference, segmented in such a way that it is distinguishable from what comes before and after (which is in general (more) flat or descending). It is categorized/scored based on gradient and length such that the score is a measure of how demanding the climb is._

For mountaineering: Are you interested in analyzing the progress of 'climbing' (how long did it take? where did I lost some time?) or only in logging that you reached the 'peak' (how many times did I climb Lagginhorn? how many peaks did I reach this month?)?

The first one is what the climb analysis is meant for (and, indeed, mainly for endurance sports), the second one is on our list of ideas (using osm data).

**Comment by Salzi** [comment]
Concerning your draft:

> A climb is a part of a route with positive elevation difference, segmented in such a way that it is distinguishable from what comes before and after **(which is in general (more) flat or descending)**. It is categorized/scored based on gradient and length such that the score is a measure of how demanding the climb is.

Is it really like that? It's what I would excepted too but when I have a look at the Lagginhorn plot it is even steeper between cl. 1 and hc. (the point with cl. 1 makes no sense to me anyway...)

Mountaineering: The main benefit of the climb score I see is in planning. When you plan a tour it really helps when you know how long it takes to climb <x> meters with <y>% in an altitude <z>. So the point "how long did it take" it's also useful for climbers. But yes there are many other parameters like partners and conditions.
On the other hand I have a few "training"-peaks which I climb several times a year and it's nice to compare them and also analyse where I lost some time.

**Comment by laufhannes** [comment]
Yes, for that description your first plot is a 'bad' example. Looking at the hc climb, the section before is more flat. That's why it's another climb section. Still, that section is steep enough to be a classified climb. The simple solution could be to merge climbs that follow each other directly, as long as both are classified.

Currently, there's one climb from 0.5 to 3.0 and a second one from 3.0 to 5.6. The merging would result in only one climb from 0.5 to 5.6, but I think the 'optimal' or at least expected solution is a climb from 1.5 (or 2.0) to 5.6. Douglas-Peucker won't choose that starting point as it looks at the line from 0.0 to 5.6 and searches for the data point which is furthest from the line (which is 3.0 where the (1) marker is located).

Any ideas for a better way to look for the starting point, assuming the peak at 5.6 is given?

**Comment by Salzi** [comment]
> ... I think the 'optimal' or at least expected solution is a climb from 1.5 (or 2.0) to 5.6...
I totally agree.

Just clicked through my dataset and there are quite a few 'bad' examples. Currently I've no better idea but I will think about.

**Comment by laufhannes** [comment]
There's another example in https://forum.runalyze.com/viewtopic.php?f=4&t=789&p=3403#p3402 which is more like the second case here: One would expect two segments/climbs, but it's selected as one.

**Comment by Salzi** [comment]
hmm. It will be very hard to develop an algorithm which fits for all...
For me personally the current algorithm works for the attached image in the forum. It would be also ok when the first part would be part of the climb but to count it as two climbs? I don't know but the first one is ~10 meters in altitude difference. To count every 10 meters as an hill would result in many climbs in some regions.
To split a climb just because the gradient changes is also strange because when you climb several 100 meters it has very rarely the same gradient.
As I said thats just my opinion. But thats shows clearly that we need a very clear definition...
